framework/src/headers/ip/v6/icmp/neighbor_options.rs

```diff
@@ -0,0 +1,206 @@
+
+    #[inline]
+    fn offset(&self) -> usize {
+        // ICMPv6 Option(Type + Length) is always 2 bytes: (8 + 8) / 8 = 2
+        2
+    }
+
+    #[inline]
+    fn size() -> usize {
+        // Icmpv6Option Header is always 4 bytes so size = offset
+        2
+    }
+
+    #[inline]
+    fn payload_size(&self, hint: usize) -> usize {
+        // There is no payload size in the ICMPv6 header
+        hint - self.offset()
+    }
+
+    #[inline]
+    fn check_correct(&self, _prev: &T) -> bool {
+        true
+    }
+}
+
+impl<T> Icmpv6Option<T>
+where
+    T: Ipv6VarHeader,
+{
+    #[inline]
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    #[inline]
+    pub fn option_type(&self) -> Option<Icmpv6OptionType> {
+        FromPrimitive::from_u8(self.option_type)
+    }
+
+    #[inline]
+    pub fn option_length(&self) -> u8 {
+        self.option_length
+    }
+}
+
+#[derive(Debug, Clone)]
+#[repr(C, packed)]
+pub struct Icmpv6RouterAdvertisementOption<T>
+where
+    T: Ipv6VarHeader,
+{
+    pub icmp_option: Icmpv6Option<T>,
+    pub source_link_layer_address: MacAddress,
+    pub _parent: PhantomData<T>,
+}
+
+impl<T> Default for Icmpv6RouterAdvertisementOption<T>
+where
+    T: Ipv6VarHeader,
+{
+    fn default() -> Icmpv6RouterAdvertisementOption<T> {
+        Icmpv6RouterAdvertisementOption {
+            icmp_option: Icmpv6Option {
+                ..Default::default()
+            },
+            source_link_layer_address: MacAddress {
+                addr: [0, 0, 0, 0, 0, 0],
+            },
+            _parent: PhantomData,
+        }
+    }
+}
+
+impl<T> fmt::Display for Icmpv6RouterAdvertisementOption<T>
+where
+    T: Ipv6VarHeader,
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "icmp_option: {} source_link_layer_address: {}",
+            self.icmp_option, self.source_link_layer_address
+        )
+    }
+}
+
+impl<T> EndOffset for Icmpv6RouterAdvertisementOption<T>
+where
+    T: Ipv6VarHeader,
+{
+    type PreviousHeader = T;
+
+    #[inline]
+    fn offset(&self) -> usize {
+        // ICMPv6 Neighbor Option(Source Link Layer Address) is always 1 byte: 8 / 8 = 1
+        1
+    }
+
+    #[inline]
+    fn size() -> usize {
+        // ICMPv6 Neighbor Option(Source Link Layer Address) is always 1 byte so size = offset
+        1
+    }
+
+    #[inline]
+    fn payload_size(&self, hint: usize) -> usize {
+        // There is no payload size in the Neighbor Option
+        hint - self.offset()
+    }
+
+    #[inline]
+    fn check_correct(&self, _prev: &T) -> bool {
+        true
+    }
+}
+
+impl<T> Icmpv6RouterAdvertisementOption<T>
+where
+    T: Ipv6VarHeader,
+{
+    #[inline]
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    #[inline]
+    pub fn source_link_layer_address(&self) -> Option<MacAddress> {
+        if self.icmp_option.option_type().unwrap() == Icmpv6OptionType::SourceLinkLayerAddress {
```

Comment:
@zeeshanlakhani @pauljamescleary according to our conversations, if we dont find the source link layer address, we would possibly return the L2 source address. I think we should probably return `None` here and let the router store the L2 source address. does the logic here seem right to you? if we don't find the Source Link Layer Address as the first option, we assume its not there

Reply:
I am still going through this.  From a type perspective, I would think we would have:

`ICMPv6NDPMessage` -- has a --> `Map<OptionType, A>` where A is a subtype of a type `NDPOption`

I am still new to Rust and Netbricks so I am unclear of if this is acceptable in either of those.

At some point, I would imagine a function that parses that map of options out.

Reply:
Lets discuss in S2 today